### ISSUE TITLE
[Feature Request] 4-op Mode planned?

### ISSUE THREAD

**Issue by bryc** [asb2m10/dexed#163]
_TL;DR: Does a 4-op compatibility mode (for editing) have a place in Dexed? Or just sysex conversion + extra waveforms on the DX7 engine?_

Sorry if this was discussed before, but I didn't see anything in the issue list. 

It would be great if Dexed could be used to load and create 4-op patches accurately. You mention wanting to import DX21/27/100 sysex and use DX11 waveforms. Instead of just converting 4-op sysex to 6-op and retrofitting Dexed's DX7 mode with extra waveforms, why not just add a full 4-op mode? That way, the imported sysex wont be just approximations, and the sound would be more accurate. I figure it could be another option in the PARM menu, which would change the parameter layout to target 4-op synths.

Without getting into the different performance settings (split/layer, MIDI, FX) the actual voice data of Yamaha 4-op synths are pretty similar to the voice data of DX7. And unlike the DX7, the 4-op synths are highly comparable across many many devices. 

By implementing the features of TX81Z/DX11, this can allow for wide compatibility. You could go beyond importing DX21/27/100 and TX81Z sysex. The voices of Yamaha YS100/YS200/V5 could be imported (barring some of their extra features). Same goes for Korg DS-8/Korg 707 which uses the same FM chip (volcafm.. hmm Korg likes stealing Yamaha's stuff!). The DX9 is also technically "compatible" but it uses the DX7's parameter ranges and has no Pitch EG or Key Velocity. 

After that, and for very little extra work, patches from Arcades, Sega MD/Genesis and NEC PC-88 could be easily converted because they use the same 4-op structure. They (probably) wont sound perfect. But the register values from [.OPM](http://vgmrips.net/wiki/OPM_File_Format), and [.VGI](https://www.plutiedev.com/format-tfi) presets are mostly compatible. 

I got a [spreadsheet](https://docs.google.com/spreadsheets/d/1k6zSwEg5iORX4E-ro_WYwr-BLq05iTB3_1uKxKwrw7s/edit?usp=sharing) of the differences between FM chips, but the main differences are:

### DX21/DX27/DX100 (YM2164 aka OPP)

Used in DX21, DX27, DX100 and the Korgs. Uses algorithms 1-8 as defined in DX7 as algorithms 1/14/8/7/5/22/31/32 (same as DX9). Operators 1&2 have to be disabled.

These are the main differences from the DX7:

- Operator **Coarse Frequency** can be set to 64 ratios instead of 32.
- Operator **Detune** is -3 to 3 instead of -7 to 7.
- LFO only has 3 waveforms instead of 5.
- EGs are simplified. Sustain level is `0-15`, Release rate is `0-15` AR/DR/SR are `0-31`. **Level cannot be set for Attack/Decay or Release**.  The **Pitch EG** also only has 3 rate/level controls, rather than 4. 
- Keyboard scaling is simplified, with just `0-3` Scaling Rate instead of `0-7` and `0-99` Level Scaling Depth. No breakpoint or curve. 
- There is no **Fine Frequency** tuning.
- No **Fixed Frequency** mode, I think? They added back in DX11.
- Some new params: **EG Bias Sensitivity (0-7)** which affects breath controller? and **AMS Enable (0-1)**.
- There is no **OSC Key Sync** (`0-1`).

Those are the main differences as far as you can tell by comparing the parameters.

## TX81Z/DX11/YS100/YS200/V50 (YM2414/YM2424 aka OPZ/OPZII)

Still 4-op, this is nearly the same as above. Except:

- Each operator can use 8 waveforms (WS param) instead of 1.
- Each patch has additional **EG Shift** param, which sets the range of envelope.
- Each patch has additional **psuedo-reverb** (slows the release rate at a certain point)
- Brings back **Fixed Frequency** mode, but differs from DX7. Has a toggle, as well as **Fixed Range** (`0-7`) and **Range Fine Tune** (`0-15`). The mark II chip used in V50 has a lower range apparently.

Also, the OPZ chips have 'performance' mode which adds some stuff that may not be relevant:

- Global FX: either Delay, Panning or Chord. Probably not important either.
- **LFO Select**. You can choose from 3 LFO settings. Seems to have two LFOs? Not sure if it's relevant.

****

As far as I know, there's no good plugin out there that can do DX21/27/100 accurately let alone TX81Z/DX11. I think it'd be a good fit for Dexed. Your thoughts?

**Comment by asb2m10** [comment]
I already tough of this. At first I was against it, but it would be doable if operator were modules and could be switch from mk1 to ym2151 (btw, the 8-bit engine is based on the YM2151). This way you can support both models. Who care if you only use 4 operator, what is important is what is operator generator is used.

It is just the sysex support that would be difficult to support *out of the box*. Needs more thinking for this.

**Comment by nsputnik** [comment]
2 op (OPL and OPL2 chips) would be great addition too.  This was used on AdLib cards, SoundBlaster cards, Yamaha Portasound series and mid 80's arcades.  I'm not sure there is a standard sysex format for this.
Regarding 4 op sysex Martin Tarenskeen has a cool DX/TX sysex conversion project but it is in python.  It's not clear what his license is but he is fairly easy to get ahold of. http://dxconvert.martintarenskeen.nl/

**Comment by bryc** [comment]
As far as I know, OPL was never used in any MIDI based synth that supported SysEx. There was PortaSound but that was basically a toy for kids.

And I hate to shut you down, but OPL instrument support is probably even less likely than DX9/DX21/DX100 instrument support, which looks like a no-go anyway (which is fine, I can get by with the various Sega Genesis VST offerings these days to achieve the same parameter ranges).

DXConvert is super useful too, you can get fairly accurate 4>6op conversions!

**Comment by nsputnik** [comment]
Sega Genesis/Megadrive used OPN chip, not OLP.  OPN is 4 operator, OLP is two operator.  There were prosumer keyboards that use OPL like Yamaha's PSR and PSS series in '87-'87.  I would not discount them just because they were used in toys.  You say you want the 4 Op lofi sound, 2 Op is even more lofi, so you should like that.  There is Plouge chipsynth PortaFM as a solution but if we are going beyond 6 Op and will also cover 4 Op then why not 2 Op?  Especially if it can be accomplished just by including additional 4 Op and 2 Op algorithms that could disable the other operators/sources/modulators.

**Comment by bryc** [comment]
Well I wouldn't get your hopes up for OPN or OPL, this issue's nearly 5 years old now. Some knight in shining armor would have to fork it and send in a PR or something, and I imagine it'd be a lot of work too. 

I appreciate OPL for what it is, since it still has part of the Yamaha FM algorithm magic. PSR seems interesting, uses the "OPQ" line of chips that I never even knew about, but it seems to be a variant of OPL2. I discounted the PortaSound stuff because the OPLL chip it uses s just a preset machine. Though generally 2-op is severely limited in what sounds you can do.

**Comment by mtarenskeen** [comment]
Not a perfect solution, but it is possible to convert patches from the 4-op synths to DX7 compatible SysEx using [DXconvert](https://martintarenskeen.nl/dxconvert) and you can many patches already converted using DXconvert in my [4OP to DX7 Conversion Project](https://martintarenskeen.nl/4opdx7)

Dsiclaimer: Results may vary
